rlaopt/expression/utils.py

```diff
@@ -2,11 +2,146 @@
+
+    # Separate constants from non-constants
+    constants = []
+    non_constants = []
+    for term in terms:
+        if isinstance(term, expr_types.constant()):
+            constants.append(term)
+        else:
+            non_constants.append(term)
+
+    # Fold all constants into one
+    if constants:
+        total = sum((c.value for c in constants[1:]), constants[0].value)
+        # Only add back if non-zero (or if it's the only term)
+        if (
+            not torch.allclose(total, torch.zeros_like(total))
+            or len(non_constants) == 0
+        ):
+            non_constants.append(expr_types.constant()(total))
+
+    # Return based on what's left
+    if len(non_constants) == 0:
+        return expr_types.constant()(0.0)
+    elif len(non_constants) == 1:
+        return non_constants[0]
+    else:
+        return expr_types.add_expr()(*non_constants)
+
+
+def _create_product(left, right, matmul: bool):
+    """Create a product expression with automatic distribution.
+
+    Distributes constants over sums to enable better splitting:
+    - const * (a + b + c) -> const*a + const*b + const*c
+    - (a + b + c) * const -> a*const + b*const + c*const
+
+    Condenses products of constants:
+    - const1 * const2 -> single const with combined value
+
+    Handles atom scaling:
+    - const * atom -> atom._scale(const.value) if atom supports it
+    - atom * const -> atom._scale(const.value) if atom supports it
+
+    Does NOT distribute expression * expression products to avoid
+    duplicating computation and increasing expression tree size.
+
+    Args:
+        left: Left operand (Expression, float, int, or torch.Tensor).
+        right: Right operand (Expression, float, int, or torch.Tensor).
+        matmul: If True, use matrix multiplication (no distribution).
+
+    Returns:
+        ProductExpression, AddExpression (if distributed), ConstExpression
+        (if both constants), or AtomExpression (if scaled atom).
+    """
+    # Convert inputs to expressions if needed
+    left = _to_expr(left)
+    right = _to_expr(right)
+
+    # Only optimize for elementwise multiplication
+    if not matmul:
+        # Check if we have constants
+        left_is_const = isinstance(left, expr_types.constant())
+        right_is_const = isinstance(right, expr_types.constant())
+
+        # Condense: const * const -> single const
+        if left_is_const and right_is_const:
+            combined_value = left.value * right.value
+            return expr_types.constant()(combined_value)
+
+        # Handle atom scaling: const * atom or atom * const
+        # Import here to avoid circular imports
+        from rlaopt.atoms.atom_expression import AtomExpression
+
+        # const * atom
+        if left_is_const and isinstance(right, AtomExpression):
+            if left.value.numel() == 1:
+                scalar = left.value.item()
+                try:
+                    return right._scale(scalar)
```

Comment:
Can we have ._scale return NotImplemented and then check for that? I think that's more ergonomic than try/catch.